### PR TITLE
meta: Add an LL1 description of the language

### PR DESCRIPTION
**Pull request by Hendiadyoin1**
This description might not be 100% accurate yet, but it is a start to
have some sort of formal documentation for the syntax.

The file name, and location dont seem ideal to me, suggestions welcome.
If I got anything wrong or forgot something, feel free to point that out

### CODE REVIEW

File: meta/language.LL1 [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+// * Multiple possible matches are separated by a vertical-bar/or-sign
+//
+// The most outer scope checks for `Declaration`s
+//
+// This is not yet as fine grained as it could be and does not differentiate
+// between some groups, so this is not (yet) a full guide to tokenization
+// Also the order of definitions and notation is currently not fully consistent
+
+// Stolen from https://webidl.spec.whatwg.org/#prod-i and adjusted to match Jakt more closely
+integer     = /-?([1-9][0-9]*|0[Xx][0-9A-Fa-f]+|0[Oo][0-7]*|0)([ui](8|16|32|64))?/
```
[the remark this answers is not in the document]
Comment (Hendiadyoin1) [reply meta/language.LL1:21]: is `z` valid as well?

File: meta/language.LL1 [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+Comment        = /\/\/.*\n/
+Import         = "import" identifier ["{" identifier {"," identifier} "}"]
+Function       = ["extern"] "function" identifier [GenericSpecifier] "(" ArgumentList ")" ["->" TypeExpression] ["throws"] [("{" {Statement} "}") | ("=>" Expression)]
+
+// NOTE: This describes Classes and Structs
+Class          = ["extern"] ("class"|"struct") identifier "{" {ClassMembers} "}"
+ClassMembers   = ClassFunction | ClassMember
+ClassMember    = [AccessModifier ["mut"]] identifier ":" TypeExpression
+ClassFunction  = [AccessModifier] "function" identifier [GenericSpecifier] "(" ArgumentList ")" ["->" TypeExpression] ["throws"] (("{" {Statement} "}") | ("=>" Expression))
+AccessModifier = "public"
```
[the remark this answers is not in the document]
Comment (Hendiadyoin1) [reply meta/language.LL1:142]: Sure missed that

File: meta/language.LL1 (hunk 1 above)
[the remark this answers is not in the document]
Comment (alimpfard) [reply meta/language.LL1]: I don't think so, `uz` is `usize`, I don't think we have `size`.

File: meta/language.LL1 [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+// The most outer scope checks for `Declaration`s
+//
+// This is not yet as fine grained as it could be and does not differentiate
+// between some groups, so this is not (yet) a full guide to tokenization
+// Also the order of definitions and notation is currently not fully consistent
+
+// Stolen from https://webidl.spec.whatwg.org/#prod-i and adjusted to match Jakt more closely
+integer     = /-?([1-9][0-9]*|0[Xx][0-9A-Fa-f]+|0[Oo][0-7]*|0)([ui](8|16|32|64)|uz)?/
+float       = /-?(([0-9]+\.[0-9]*|[0-9]*\.[0-9]+)([Ee][+-]?[0-9]+)?|[0-9]+[Ee][+-]?[0-9]+)(f64)?/
+            | /-?([1-9][0-9]*|0[Xx][0-9A-Fa-f]+|0[Oo][0-7]*|0)f64/
```
Comment (AtkinsSJ) [inline meta/language.LL1:23]: f32 is also a suffix: https://github.com/SerenityOS/jakt/blob/a0fd770771c134149f8b11605d295280207eb519/src/lexer.rs#L583

I think you asked about `z`, and that's not listed. (I thought it was one but I don't have the best memory.)

Comment (Hendiadyoin1) [reply meta/language.LL1]: Thanks
Added that aswell

File: meta/language.LL1 [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+// NOTE: This describes Classes and Structs
+Class          = ["extern"] ("class"|"struct") identifier "{" {ClassMembers} "}"
+ClassMembers   = ClassFunction | ClassMember
+ClassMember    = [AccessModifier ["mut"]] identifier ":" TypeExpression
+ClassFunction  = [AccessModifier] "function" identifier [GenericSpecifier] "(" ArgumentList ")" ["->" TypeExpression] ["throws"] (("{" {Statement} "}") | ("=>" Expression))
+AccessModifier = "public"
+               | "private"
+               | "restricted" "(" identifier [GenericSpecifier] {"," identifier [GenericSpecifier]} ")"
+
+Enum           = ["boxed"] "enum" identifier [GenericSpecifier] [":" TypeExpression] "{" EnumMember {EnumMember} "}"
```
Comment (alimpfard) [inline meta/language.LL1:175]: `GenericSpecifier` and `":" TypeExpression` are mutually exclusive (and so is `"boxed"` with `":" TypeExpression`).

File: meta/language.LL1 [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+EnumMember     = ClassFunction
+               | identifier ["(" [identifier ":"] TypeExpression {"," [identifier ":"] TypeExpression} ")"]
```
Comment (alimpfard) [inline meta/language.LL1:177]: Given an enum with an underlying type, the members can also be `identifier ["=" Expression]`.

File: meta/language.LL1 [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+Parameter      = [identifier ":"] Expression
+
+Declaration    = Comment
+               | Namespace
+               | Import
+               | Function
+               | Class
+               | Enum
+
+Comment        = /\/\/.*\n/
```
Comment (alimpfard) [inline meta/language.LL1:161]: This won't match comments at the end of the file.

```suggestion
Comment        = /\/\/.*$/
```

File: meta/language.LL1 [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+IfStatement          = "if" BooleanExpression BlockStatement ["else" BlockStatement]
+TryStatement         = "try" (Expression | BlockStatement) "catch" identifier BlockStatement
+ForStatement         = "for" identifier "in" Expression BlockStatement
+LoopStatement        = "loop" BlockStatement
+WhileStatement       = "while" Expression BlockStatement
+
+// NOTE: `mut this` is only allowed in ClassFunctions
+ArgumentList   = [[["mut"] "this"] [Arguments]]
+Arguments      = "," Argument [Arguments]
```
Comment (alimpfard) [inline meta/language.LL1:147]: This allows `foo(, bar: T)`

File: meta/language.LL1 [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+Argument       = ["anon"] ["mut"] identifier [":" TypeExpression]
```
Comment (alimpfard) [inline meta/language.LL1:148]: I'm pretty sure the type annotation is required here

File: meta/language.LL1 [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+BlockStatement       = "{" Statement "}"
+
+IfStatement          = "if" BooleanExpression BlockStatement ["else" BlockStatement]
+TryStatement         = "try" (Expression | BlockStatement) "catch" identifier BlockStatement
+ForStatement         = "for" identifier "in" Expression BlockStatement
+LoopStatement        = "loop" BlockStatement
+WhileStatement       = "while" Expression BlockStatement
+
+// NOTE: `mut this` is only allowed in ClassFunctions
+ArgumentList   = [[["mut"] "this"] [Arguments]]
```
Comment (alimpfard) [inline meta/language.LL1:146]: This is a Parameter...

File: meta/language.LL1 [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+Arguments      = "," Argument [Arguments]
+Argument       = ["anon"] ["mut"] identifier [":" TypeExpression]
+
+ParameterList  = [Parameter [Parameters]]
```
Comment (alimpfard) [inline meta/language.LL1:150]: ...and this is an Argument.

File: meta/language.LL1 [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+// Also the order of definitions and notation is currently not fully consistent
+
+// Stolen from https://webidl.spec.whatwg.org/#prod-i and adjusted to match Jakt more closely
+integer     = /-?([1-9][0-9]*|0[Xx][0-9A-Fa-f]+|0[Oo][0-7]*|0)([ui](8|16|32|64)|uz)?/
+float       = /-?(([0-9]+\.[0-9]*|[0-9]*\.[0-9]+)([Ee][+-]?[0-9]+)?|[0-9]+[Ee][+-]?[0-9]+)(f(64|32))?/
+            | /-?([1-9][0-9]*|0)f(64|32)/
+identifier  = /_?[A-Za-z][0-9A-Z_a-z-]*/
+
+String      = '"' EscapedContent '"'
+ByteString = "b'" EscapedContent "'"
```
Comment (alimpfard) [inline meta/language.LL1:27]: You mean a ByteLiteral? this thing can only contain a single byte in the ASCII range afaik

File: meta/language.LL1 [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+MatchCase        = MatchKey "=>" Statement
+MatchKey         = MatchKey "|" MatchKey
+                 | {identifier "::"} identifier ["(" {identifier [":" identifier]} ")"]
+                 | Constant
+                 | "else"
+
+IsExpression     = Expression "is" TypeExpression
+CastExpression   = Expression "as!" TypeExpression
+
+Expression       = Expression op Expression
```
Comment (alimpfard) [inline meta/language.LL1:97]: `Expression "is" Expression` is not a valid construct (op => boolean_op => "is").

File: meta/language.LL1 [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+                 | lop Expression
+                 | Expression rop
```
Comment (alimpfard) [inline meta/language.LL1:99]: Left recursion isn't valid in LL(k).

File: meta/language.LL1 [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+                 | "(" Expression ")"
+                 | Constant
+                 | ArrayLiteral
+                 | MapLiteral
+                 | SetLiteral
+                 | Range
```
Comment (alimpfard) [inline meta/language.LL1:105]: This technically can't be here as it causes left recursion (there's tonnes of this, I won't bother with the rest).

File: meta/language.LL1 [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+
+MatchExpression  = "match" Expression "{" {MatchCase} "}"
+MatchCase        = MatchKey "=>" Statement
+MatchKey         = MatchKey "|" MatchKey
+                 | {identifier "::"} identifier ["(" {identifier [":" identifier]} ")"]
+                 | Constant
+                 | "else"
+
+IsExpression     = Expression "is" TypeExpression
+CastExpression   = Expression "as!" TypeExpression
```
Comment (alimpfard) [inline meta/language.LL1:95]: There's also `as?`.

File: meta/language.LL1 [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+Comment        = /\/\/.*\n/
+Namespace      = "namespace" identifier "{" {Declaration} "}"
+Import         = "import" identifier ["{" identifier {"," identifier} "}"]
+Function       = ["extern"] "function" identifier [GenericSpecifier] "(" ArgumentList ")" ["->" TypeExpression] ["throws"] [("{" {Statement} "}") | ("=>" Expression)]
+
+// NOTE: This describes Classes and Structs
+Class          = ["extern"] ("class"|"struct") identifier "{" {ClassMembers} "}"
+ClassMembers   = ClassFunction | ClassMember
+ClassMember    = [AccessModifier ["mut"]] identifier ":" TypeExpression
+ClassFunction  = [AccessModifier] "function" identifier [GenericSpecifier] "(" ArgumentList ")" ["->" TypeExpression] ["throws"] (("{" {Statement} "}") | ("=>" Expression))
```
Comment (alimpfard) [inline meta/language.LL1:170]: A function (class or not) is allowed to have no body.

File: meta/language.LL1 [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+AccessModifier = "public"
+               | "private"
+               | "restricted" "(" identifier [GenericSpecifier] {"," identifier [GenericSpecifier]} ")"
+
+Enum           = ["boxed"] "enum" identifier [GenericSpecifier] [":" TypeExpression] "{" EnumMember {EnumMember} "}"
+EnumMember     = ClassFunction
```
Comment (alimpfard) [inline meta/language.LL1:176]: This currently doesn't accept `restricted(...)` (but it should in the future).

**Review by alimpfard — COMMENTED** [review]
Sorry I took a while to review this ^^'
The left recursions can be left alone since this isn't really trying to conform to either ebnf or ll(k), but I figured I'd keep the comments.

File: meta/language.LL1 (hunk 12 above)
Comment (Hendiadyoin1) [reply meta/language.LL1:97]: Ah ok
Wa confused with the variant-is, but the IsExpression should magically handle that one aswell

File: meta/language.LL1 (hunk 15 above)
Comment (Hendiadyoin1) [reply meta/language.LL1:95]: I could not find a user of that in the current code base, so I was not sure if that exists

File: meta/language.LL1 (hunk 17 above)
Comment (Hendiadyoin1) [reply meta/language.LL1:176]: wdym, an ellipsis in the restriction?

File: meta/language.LL1 (hunk 13 above)
Comment (Hendiadyoin1) [reply meta/language.LL1:99]: Ok I currently dont really know how to fix that one, but I will add a note to the introduction

File: meta/language.LL1 (hunk 17 above)
Comment (alimpfard) [reply meta/language.LL1:176]: No, I mean restricted (anything)
Currently only public/private is accepted there iirc

File: meta/language.LL1 [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+                 | ArrayLiteral
+                 | MapLiteral
+                 | SetLiteral
+                 | Range
+                 | MatchExpression
+                 | IsExpression
+                 | CastExpression
+                 | identifier {"::" identifier}
+                 | [Expression] /!?\./ identifier
+                 | Expression "(" ArgumentList ")"
```
Comment (Hendiadyoin1) [inline meta/language.LL1:111]: I feel there are some missing `[GenericSpecifier]`s missing here
@alimpfard can you confirm that any identifier here can take a GenericSpecifier?

Comment (alimpfard) [reply meta/language.LL1:111]: Only the call target can atm
```
Expression ["<" {Typename} ">"] "(" ... ")"
```

Comment (Hendiadyoin1) [reply meta/language.LL1:111]: So something like `my_enum<i64>::myvalue` is not allowed currently?
Same for any of
```
MyClass<T>::static_function(...)
MyClass::static_function<T>(...)
generic_constant<T>
```
My guess would be that the GenericSpecifier belongs to the identifier and not to the Expression

Comment (alimpfard) [reply meta/language.LL1:111]: Yep, that's currently not valid.

**Comment by cybergsus** [comment]
You can't iterate over a Range<Range<_>> but, even if it sounds goofy, the syntax allows it: `(0..2)..(8..3)`. This makes no logical sense, but the parser can be made simpler and then reject any iteration that goes over that type

**Comment by mnlrsn** [comment]
> 

> even if it sounds goofy, the syntax allows it

Sure, that was my point, too. The question is more if it *should* allow it...

**Comment by cybergsus** [comment]
> Sure, that was my point, too. The question is more if it *should* allow it...

I don't feel like prohibiting this explicitly syntax-wise, since it's nature is more specific to the **type** *inside* of the range. That's OK to be parsed, a range is an expression. It will be rejected by typechecker. Disallowing this pattern explicitly in syntax only adds noise to the language spec and to the parser.

**Pull request closed without merging**